src/audio/drc/drc.c

```diff
@@ -0,0 +1,849 @@
+{
+	int64_t num;
+	int32_t x2; /* Q12.20 */
+	int32_t x_db, x2_db, y_db, y2_db; /* Q8.24 */
+
+	if (x < state->linear_threshold)
+		return Q_CONVERT_FLOAT(1.0, 12);
+
+	x2 = Q_MULTSR_32X32((int64_t)x, Q_CONVERT_FLOAT(1.001, 12), 12, 12, 12);
+
+	x_db = lin2db_fixed(x);
+	x2_db = lin2db_fixed(x2);
+
+	y_db = lin2db_fixed(knee_curve(state, x, k));
+	y2_db = lin2db_fixed(knee_curve(state, x2, k));
+
+	num = Q_SHIFT_LEFT((int64_t)(y2_db - y_db), 24, 36);
+	/* Q20.12 = Q28.36 / Q8.24 */
+	return (int32_t)(num / (x2_db - x_db));
+}
+
+/* Input: Q20.12 -> Output: Q20.12 */
+static int32_t k_at_slope(struct drc_state *state, int32_t desired_slope)
+{
+	int32_t x_db;  /* Q8.24 */
+	int32_t x;     /* Q12.20 */
+	int32_t slope; /* Q20.12 */
+
+	int32_t min_k, max_k, k; /* Q20.12 */
+	int i;
+
+	x_db = state->db_threshold + state->db_knee;
+	x = db2lin_fixed(x_db);
+
+	/* Approximate k given initial values */
+	min_k = Q_CONVERT_FLOAT(0.1, 12);
+	max_k = Q_CONVERT_FLOAT(10000.0, 12);
+	k = Q_CONVERT_FLOAT(5.0, 12);
+
+	for (i = 0; i < 15; ++i) {
+		/* A high value for k will more quickly asymptotically approach
+		 * a slope of 0. */
+		slope = slope_at(state, x, k);
+		if (slope < desired_slope) {
+			/* k is too high */
+			max_k = k;
+		} else {
+			/* k is too low */
+			min_k = k;
+		}
+
+		/* Re-calculate based on geometric mean */
+		k = Q_CONVERT_FLOAT(sqrtf(Q_CONVERT_QTOF(min_k, 12) * Q_CONVERT_QTOF(max_k, 12)), 12);
+	}
+
+	return k;
+}
+
+static inline void update_static_curve_parameters(struct drc_state *state,
+						  int32_t *params)
+{
+	int64_t num;
+	int32_t k;
+	int32_t y0;
+
+	if (state->db_threshold == params[0] && state->db_knee == params[1] &&
+	    state->ratio == params[2])
+		return;
+
+	/* Update threshold and knee */
+	state->db_threshold = params[0];
+	state->linear_threshold = db2lin_fixed(state->db_threshold);
+	state->db_knee = params[1];
+
+	/* Compute knee parameters */
+	state->ratio = params[2];
+	num = Q_SHIFT_LEFT((int64_t)DRC_ONE_Q30, 30, 44);
+	/* Q12.20 = Q20.44 / Q8.24 */
+	state->slope = (int32_t)(num / state->ratio);
+	state->K = k_at_slope(state, state->slope);
+
+	/* Parameters for knee_curveK() usage */
+	num = Q_SHIFT_LEFT((int64_t)DRC_ONE_Q30, 30, 32);
+	state->knee_alpha = dk->threshold + (int32_t)(num / state->K);
+	num = Q_SHIFT_LEFT((int64_t)Q_CONVERT_QTOF(-expf(Q_CONVERT_FLOAT(Q_MULTSR_32X32((int64_t)state->K, state->linear_threshold, 12, 20, 30))), 30), 42);
+	/* Q2.30 = Q22.42 / Q20.12 */
+	state->knee_beta = (int32_t)(num / state->K);
+	state->knee_threshold = db2lin_fixed(state->knee_threshold + state->db_knee);
+
+	/* Parameters for volume_gain() usage */
+	y0 = knee_curve(state, state->knee_threshold, state->K); /* Q12.20 */
+	state->ratio_base = Q_CONVERT_FLOAT(Q_CONVERT_QTOF(y0, 20) * powf(Q_CONVERT_QTOF(state->knee_threshold, 20), Q_CONVERT_QTOF(-state->slope, 20)), 30);
+}
+
+static inline void drc_set_parameters(struct drc_state *state,
```

Comment:
this looks really heavy and could cause glitches, is this done once per config? if so we should do this on the cpu and pass the calculated coefficients

Reply:
This makes senses. After more discussion and consideration, I would like to re-organize the DRC component with the following steps:
1. Because drc_set_parameters are heavy and done once per config, we should do this on the cpu and pass the calculated coefficients from sof_drc_config (coefficients are not required to be copied).
2. Implement DRC frame processing code in drc_generic.c
3. drc_state should only stores register-like variables which may be initialized by new stream, loaded and saved between frame or division processing.

Reply:
@singalsu any inputs ?

Reply:
@lgirdwood this is how all the processing components work including eq. I think we are good here.